### ISSUE TITLE
Login to external cluster does not happen

### ISSUE THREAD

**Issue by michaelsauter** [opendevstack/ods-mro-jenkins-shared-library#199]
Even when an external API is configured, the login does not happen in Jenkins 2.176.4.

Quiz time - what does the following method return?
https://github.com/opendevstack/ods-mro-jenkins-shared-library/blob/2.x/src/org/ods/util/Project.groovy#L450-L464
```groovy
@NonCPS
boolean getTargetClusterIsExternal() {
    def isExternal = false
    def sessionApiUrl = this.data.openshift.sessionApiUrl
    def targetApiUrl = this.data.openshift.targetApiUrl
    def targetApiUrlMatcher = targetApiUrl =~ /:[0-9]+$/
    if (targetApiUrlMatcher.find()) {
        isExternal = sessionApiUrl != targetApiUrl
    } else {
        def sessionApiUrlWithoutPort = sessionApiUrl.split(':').dropRight(1).join(':')
        isExternal = sessionApiUrlWithoutPort != targetApiUrl
    }
    this.steps.echo("Cluster ${targetApiUrl} is external=${isExternal}")
    isExternal
}
```

Given the signature, one could expect `false` or `true`. Just looking at the method body, one should also assume that it can only be `false` or `true`.  But Jenkins, in version 2.176.4, returns ... drumroll ... `null`! 😂 

Some googling lead to https://issues.jenkins-ci.org/browse/JENKINS-26481?focusedCommentId=371870&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-371870.

In short, calling a method NOT annotated with `@NonCPS` from a method annotated with `@NonCPS` returns the value from that method call (returning early from the method). In our example, `getTargetClusterIsExternal` returns the value returned from `this.steps.echo()`, which is `null`. The bug does not seem to happen in newer Jenkins versions (e.g. 2.204), but I could not find which version fixes it.

In order to allow the MRO library to work with earlier versions of Jenkins, I'm going to remove the echo statement from `getTargetClusterIsExternal`.

FYI @borja44

**Comment by martsec** [comment]
Good catch!

**Comment by oalyman** [comment]
@michaelsauter I think this should work as long as this.step.echo always returns null, right?:
`this.steps.echo("Cluster ${targetApiUrl} is external=${isExternal}") || isExternal`

**Comment by metmajer** [comment]
> Given the signature, one could expect false or true. Just looking at the method body, one should also assume that it can only be false or true. But Jenkins, in version 2.176.4, returns ... drumroll ... null! 😂

@michaelsauter Wow, that's a bummer. I've learned the hard way myself that Jenkins jobs in `/vars` always return type `Object`, but a `@NonCPS` annotated method returning `null` instead of a boolean value is really hard stuff. Jenkins...

Mind that you're calling `this.steps.echo` in a `@NonCPS` annotated method, although all methods in the Jenkins context (aka. `steps`) use CPS. Doing so leads to undefined behavior, as stated in https://www.jenkins.io/doc/book/pipeline/cps-method-mismatches/#use-of-pipeline-steps-fromnoncps. FYI @martsec @jorge-romero
